stock_dynamic_routing_delivery/tests/test_carrier_propagation.py

```diff
@@ -0,0 +1,164 @@
+        # Disable carrier propagation on the pick rule
+        # We want the one in the dynamic routing to play out
+        cls.warehouse.delivery_route_id.rule_ids.filtered(
+            lambda r: r.picking_type_id == cls.warehouse.pick_type_id
+        ).propagate_carrier = False
+        # Set up the products
+        cls.product1 = cls.env["product.product"].create(
+            {"name": "Product 1", "type": "consu", "is_storable": True}
+        )
+        cls.product2 = cls.env["product.product"].create(
+            {"name": "Product 2", "type": "consu", "is_storable": True}
+        )
+        # Set up the locations
+        cls.stock_location = cls.warehouse.lot_stock_id
+        cls.location_hb = cls.env["stock.location"].create(
+            {"name": "Highbay", "location_id": cls.stock_location.id}
+        )
+        cls.location_handover = cls.env["stock.location"].create(
+            {"name": "Handover", "location_id": cls.stock_location.id}
+        )
+        # Set up some stocks
+        cls._update_product_qty_in_location(cls.location_hb, cls.product1, 100)
+        cls._update_product_qty_in_location(cls.stock_location, cls.product2, 100)
+        # Set up the dynamic routing
+        cls.pick_type_routing = cls.env["stock.picking.type"].create(
+            {
+                "name": "Dynamic Routing",
+                "code": "internal",
+                "sequence_code": "WH/HO",
+                "warehouse_id": cls.warehouse.id,
+                "use_create_lots": False,
+                "use_existing_lots": True,
+                "default_location_src_id": cls.location_hb.id,
+                "default_location_dest_id": cls.location_handover.id,
+                "reservation_method": "manual",
+            }
+        )
+        cls.routing = cls.env["stock.routing"].create(
+            {
+                "location_id": cls.location_hb.id,
+                "picking_type_id": cls.warehouse.pick_type_id.id,
+                "rule_ids": [
+                    Command.create(
+                        {
+                            "method": "pull",
+                            "picking_type_id": cls.pick_type_routing.id,
+                            "propagate_carrier": True,
+                        }
+                    ),
+                ],
+            }
+        )
+        # Set up the test order
+        cls.carrier_product = cls.env["product.product"].create(
+            {
+                "type": "service",
+                "name": "Shipping costs",
+                "standard_price": 10,
+                "list_price": 100,
+            }
+        )
+        cls.carrier = cls.env["delivery.carrier"].create(
+            {
+                "name": "Test carrier",
+                "delivery_type": "fixed",
+                "fixed_price": 50,
+                "product_id": cls.carrier_product.id,
+            }
+        )
+        cls.order = cls.env["sale.order"].create(
```

Comment:
Do not depend on sale if possible.

Reply:
We have to depend on `stock_delivery`, which ultimately depends on `sale`, because that's where the field is defined:
https://github.com/odoo/odoo/blob/9203b3631f4a200bef5e627597e78eada521af7d/addons/stock_delivery/models/stock_picking.py#L23